csrc/ops/arith.cpp

```diff
@@ -2684,4 +2684,57 @@ TensorView* tensor(Val* val) {
   return out;
 }
 
+namespace {
+
+//! Create new output for matmul
+static TensorView* newForMatmul(
+    TensorView* tv_a,
+    TensorView* tv_b
+) {
+  auto orig_domain_a = TensorDomain::noReductions(tv_a->getMaybeRFactorDomain());
+  auto orig_domain_b = TensorDomain::noReductions(tv_b->getMaybeRFactorDomain());
+  auto ndims_a = orig_domain_a.size();
+  auto ndims_b = orig_domain_b.size();
+
+  NVF_ERROR(ndims_a>=1 && ndims_b>= 1);
+
+  std::vector<IterDomain*> new_domain;
+  auto higher_dim_domain = ndims_a >= ndims_b ? orig_domain_a : orig_domain_b;  
+  new_domain.reserve(higher_dim_domain.size());
+
+  if (ndims_a > 2 || ndims_b > 2) {
+    for (auto inx: c10::irange(higher_dim_domain.size() - 2)) {
+      new_domain.push_back(IterDomainBuilder(higher_dim_domain[inx]).resetSchedulingParams().build());
```

Comment:
Q: We need to check broadcastability of inputs. I was doing it using a check on the extents -- either one of the extents is 1 or the extents are the same. This does not work for symbolic inputs? Do we rely on thunder to reject input sizes that are incompatible?

Reply:
The output IterDomain should be symbolic in that case. Dynamic fusion concretization should set the IterType automatically. This should work pretty much just like how a binary op works.

Reply:
Specifically, we usually use this function to create a new output domain for things like binary ops: https://github.com/NVIDIA/Fuser/blob/378a8386b082456666b3949349ac3640ed7db5ff/csrc/ops/utils.cpp#L183.
I think we should refactor this to create a utility that takes a vector of aligned `IterDomain`s and produces a new output `IterDomain`. `newOutputDomain` would create one of these vectors for each dimension using values from each `TensorView` input. In your case you would call the new function directly with the pair of corresponding axes.

Reply:
If this PR is working with concrete inputs now but not symbolic, then we can do that refactor in a separate PR. We should just guard the `matmul` function with a check that none of the input axes `isSymbolic()` until that is addressed/tested.

Reply:
> Specifically, we usually use this function to create a new output domain for things like binary ops:
> 
> https://github.com/NVIDIA/Fuser/blob/378a8386b082456666b3949349ac3640ed7db5ff/csrc/ops/utils.cpp#L183
> 
> .
> I think we should refactor this to create a utility that takes a vector of aligned `IterDomain`s and produces a new output `IterDomain`. `newOutputDomain` would create one of these vectors for each dimension using values from each `TensorView` input. In your case you would call the new function directly with the pair of corresponding axes.

I am not clear on this. Could you give me a small example on what the inputs will be?

Reply:
Do you mean something like `IterDomain* newIterDomain (const IterDomain* id)`?

Reply:
I think I got your point. Will ping you once I finish the refactor.

Reply:
Currently, `newOutputDomain` takes a vector of `TensorView`s and assumes the axes are aligned in the trivial way, but that will need to be more flexible for your case. So I think you'd want something like `IterDomain* newIterDomain(const std::vector<IterDomain*> ids)` and the argument would be all of the mapped producer IDs.